File: app/src/main/java/com/bcg/gpscompass/ui/screen/location/LocationFragment.java
//package com.bcg.gpscompass.ui.screen.location;
//
//import android.Manifest;
//import android.content.pm.PackageManager;
//import android.os.Bundle;
//import android.view.LayoutInflater;
//import android.view.View;
//import android.view.ViewGroup;
//
//import androidx.activity.result.ActivityResultLauncher;
//import androidx.activity.result.contract.ActivityResultContracts;
//import androidx.annotation.NonNull;
//import androidx.annotation.Nullable;
//import androidx.appcompat.widget.AppCompatTextView;
//import androidx.core.content.ContextCompat;
//import androidx.fragment.app.Fragment;
//
//import com.bcg.gpscompass.R;
//import com.mapbox.maps.MapView;
//import com.mapbox.maps.Style;
//import com.mapbox.maps.plugin.Plugin;
//import com.mapbox.maps.plugin.locationcomponent.LocationComponentPlugin;
//import com.mapbox.maps.plugin.locationcomponent.generated.LocationComponentSettings;
//
//import kotlin.Unit;
//import kotlin.jvm.functions.Function1;
//
//public class LocationFragment extends Fragment {
//
//    private final ActivityResultLauncher<String> requestPermissionLauncher = registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
//        if (isGranted) {
//            //TODO progress
//        } else {
//            // Reject
//        }
//    });
//    private MapView mMapView;
//    private AppCompatTextView mTvAddress;
//    private AppCompatTextView mTvLongitude;
//    private AppCompatTextView mTvLatitude;
//
//    public LocationFragment() {
//    }
//
//    public static LocationFragment newInstance() {
//        LocationFragment fragment = new LocationFragment();
//        Bundle args = new Bundle();
//        fragment.setArguments(args);
//        return fragment;
//    }
//
//    @Override
//    public void onCreate(Bundle savedInstanceState) {
//        super.onCreate(savedInstanceState);
//        if (getArguments() != null) {
//        }
//    }
//
//    @Override
//    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
//        // Inflate the layout for this fragment
//        return inflater.inflate(R.layout.fragment_location, container, false);
//    }
//
//    @Override
//    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
//        super.onViewCreated(view, savedInstanceState);
//        mMapView = view.findViewById(R.id.mapView);
//        mTvAddress = view.findViewById(R.id.tv_address);
//        mTvLatitude = view.findViewById(R.id.tv_latitude);
//        mTvLongitude = view.findViewById(R.id.tv_longitude);
//        mMapView.getMapboxMap().loadStyleUri(Style.MAPBOX_STREETS, new Style.OnStyleLoaded() {
//            @Override
//            public void onStyleLoaded(@NonNull Style style) {
//                LocationComponentPlugin componentPlugin = mMapView.getPlugin(Plugin.MAPBOX_LOCATION_COMPONENT_PLUGIN_ID);
//                componentPlugin.updateSettings(new Function1<LocationComponentSettings, Unit>() {
//                    @Override
//                    public Unit invoke(LocationComponentSettings locationComponentSettings) {
////                        locationComponentSettings.setEnabled();
//                        return null;
//                    }
//                });
//            }
//        });
//    }
//
//
//    private void enableLocationComponent(@NonNull Style loadedMapStyle) {
//        if (ContextCompat.checkSelfPermission(requireActivity(), Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
//            LocationComponentOptions customLocationComponentOptions = LocationComponentOptions.builder(requireActivity()).pulseEnabled(true).build();
//            LocationComponent locationComponent = mapboxMap.getLocationComponent();
//            locationComponent.activateLocationComponent(LocationComponentActivationOptions.builder(requireActivity(), loadedMapStyle).locationComponentOptions(customLocationComponentOptions).build());
//            locationComponent.setLocationComponentEnabled(true);
//            locationComponent.setCameraMode(CameraMode.TRACKING);
//            locationComponent.setRenderMode(RenderMode.NORMAL);
//        } else if (shouldShowRequestPermissionRationale("ABC")) {
//            //showInContextUI(...);
//        } else {
//            requestPermissionLauncher.launch(Manifest.permission.ACCESS_FINE_LOCATION);
//        }
//    }

File: app/src/main/java/com/bcg/gpscompass/ui/screen/location/LocationFragment.kt
package com.bcg.gpscompass.ui.screen.location

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.AppCompatTextView
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import com.bcg.gpscompass.R
import com.mapbox.geojson.Point
import com.mapbox.maps.CameraOptions
import com.mapbox.maps.EdgeInsets
import com.mapbox.maps.MapView
import com.mapbox.maps.Style.Companion.MAPBOX_STREETS
import com.mapbox.maps.plugin.animation.MapAnimationOptions
import com.mapbox.maps.plugin.animation.camera
import com.mapbox.maps.plugin.locationcomponent.location


private const val ARG_ADDRESS = "address"
private const val ARG_LATITUDE = "latitude"
private const val ARG_LONGITUDE = "longitude"

class LocationFragment : Fragment() {
    private var address: String? = null
    private var latitude: Double? = null
    private var longitude: Double? = null
    private lateinit var mToolbar: Toolbar
    private lateinit var mMapView: MapView
    private lateinit var mTvAddress: AppCompatTextView
    private lateinit var mTvLongitude: AppCompatTextView
    private lateinit var mTvLatitude: AppCompatTextView
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            address = it.getString(ARG_ADDRESS)
            latitude = it.getDouble(ARG_LATITUDE)
            longitude = it.getDouble(ARG_LONGITUDE)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_location, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mMapView = view.findViewById(R.id.mapView)
        mToolbar = view.findViewById(R.id.toolBar)
        mToolbar.setNavigationIcon(R.drawable.ic_back_white)
        mToolbar.setNavigationOnClickListener(View.OnClickListener { activity!!.onBackPressedDispatcher.onBackPressed() })
        mTvAddress = view.findViewById(R.id.tv_address)
        mTvLatitude = view.findViewById(R.id.tv_latitude)
        mTvLongitude = view.findViewById(R.id.tv_longitude)
        mMapView.getMapboxMap().loadStyleUri(MAPBOX_STREETS) {
            mMapView.location.updateSettings {
                enabled = true
                pulsingEnabled = true
            }
        }
        mTvAddress.text = address ?: ""
        mTvLatitude.text = latitude?.toString()
        mTvLongitude.text = longitude?.toString()
        initCamera();
    }

    private fun initCamera() {
        val cameraPosition = CameraOptions.Builder()
            .zoom(12.0)
            .center(Point.fromLngLat(longitude ?: 0.0, latitude ?: 0.0))
            .build()
        mMapView.getMapboxMap().setCamera(cameraPosition)
    }

    override fun onStart() {
        super.onStart()
        mMapView!!.onStart()
    }

    override fun onStop() {
        super.onStop()
        mMapView!!.onStop()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mMapView!!.onLowMemory()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mMapView!!.onDestroy()
    }

    companion object {
        fun newInstance(address: String?, latitude: Double, longitude: Double) =
            LocationFragment().apply {
                arguments = Bundle().apply {
                    putString(ARG_ADDRESS, address)
                    putDouble(ARG_LATITUDE, latitude)
                    putDouble(ARG_LONGITUDE, longitude)
                }
            }
    }
}